src/Sparrow/Utils/ZstdStream.cs

```diff
@@ -196,56 +198,113 @@ public override void Flush()
         {
             using (_disposerLock.EnsureNotDisposed())
             {
-                while (true)
-                {
-                    var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
-                    if (outputBytes == 0)
-                        break;
-                    _inner.Write(_tempBuffer, 0, outputBytes);
-                }
-                _inner?.Flush();
+                FlushInternal();
             }
         }
 
+        private void FlushInternal()
+        {
+            while (true)
+            {
+                var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
+                if (outputBytes == 0)
+                    break;
+                _inner.Write(_tempBuffer, 0, outputBytes);
+            }
+
+            _inner.Flush();
+        }
+
         public override async Task FlushAsync(CancellationToken cancellationToken)
         {
             using (await _disposerLock.EnsureNotDisposedAsync().ConfigureAwait(false))
             {
-                while (true)
-                {
-                    var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
-                    if (outputBytes == 0)
-                        break;
+                await FlushInternalAsync(cancellationToken).ConfigureAwait(false);
+            }
+        }
 
-                    await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
-                }
-                await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
+        private async Task FlushInternalAsync(CancellationToken cancellationToken = default)
+        {
+            while (true)
+            {
+                var (outputBytes, _, _) = CompressStep(ReadOnlySpan<byte>.Empty, ZstdLib.ZSTD_EndDirective.ZSTD_e_flush);
+                if (outputBytes == 0)
+                    break;
+
+                await _inner.WriteAsync(_tempBuffer, 0, outputBytes, cancellationToken).ConfigureAwait(false);
             }
+
+            await _inner.FlushAsync(cancellationToken).ConfigureAwait(false);
         }
 
-        private void DisposeInternal()
+        public override async ValueTask DisposeAsync()
         {
+            if (_disposed)
```

Comment:
I would expect that `_disposerLock` will throw if we double dispose... 
BTW, why do we have here a double dispose? is that expected?